pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -334,21 +335,35 @@ func conditionallyConvertToIstioMtls(tls *networking.TLSSettings, serviceAccount
 		if len(sniToUse) == 0 {
 			sniToUse = sni
 		}
-		return buildIstioMutualTLS(serviceAccounts, sniToUse)
+		return buildIstioMutualTLS(serviceAccounts, trustDomain, sniToUse)
 	}
 	return tls
 }
 
 // buildIstioMutualTLS returns a `TLSSettings` for ISTIO_MUTUAL mode.
-func buildIstioMutualTLS(serviceAccounts []string, sni string) *networking.TLSSettings {
+func buildIstioMutualTLS(serviceAccounts []string, trustDomain, sni string) *networking.TLSSettings {
+	var accounts []string
+	accounts = append(accounts, serviceAccounts...)
+
+	// When trust domain is set, and the service account is a k8s service account with format like 'spiffe://cluster.local/ns/foo/sa/bar',
+	// add account with format like 'spiffe://trustDomain/ns/foo/sa/bar' to SubjectAltNames, which will be used for secure naming.
+	// Put two formats for now until switch is done.
+	if trustDomain != "" {
+		for _, sa := range serviceAccounts {
+			ac := strings.Replace(sa, "cluster.local", trustDomain, -1)
+			accounts = append(accounts, ac)
+		}
+	}
```

Comment:
Your branch doesn't seem to have the right code.. 
https://github.com/istio/istio/pull/10215/files#diff-f507fb55b7edf4a03d3c7e9aebf64a1eR209 serviceAccounts is always nil. 
Second, please move this logic to GetIstioServiceAccounts(hostname, port, trustDomain).

Reply:
1. my branch is synced with latest master, in which ```applyTrafficPolicy``` pass nil as ```serviceAccounts``` in ```buildOutboundSniDnatClusters``` (but not ```buildOutboundClusters```)
2. I thought about that, the reason I kept in cluster.go is because this identity format is only for k8s, while GetIstioServiceAccounts is cross-platform interface